### PR TITLE
refactor: use neverthrow in /:transactionId/otp (part 5)

### PR DESCRIPTION
**Pull request by mantariksh**
Builds on #1453

## Problem

See #1450

## Solution

This is the 5th part of the solution, which focuses on the `POST /transaction/:transactionId/otp` endpoint.

## Tests
See #1450

### CODE REVIEW

File: src/app/modules/verification/verification.controller.ts [hunk 1]
```diff
@@ -118,28 +119,40 @@ export const handleResetField: RequestHandler<
  * @param req
  * @param res
  */
-export const getNewOtp: RequestHandler<
+export const handleGetOtp: RequestHandler<
   { transactionId: string },
-  string,
+  { message: string },
```
Comment (karrui) [inline src/app/modules/verification/verification.controller.ts:124]: nit: use `ErrorDto` type instead?

File: src/app/modules/verification/verification.service.ts [hunk 2]
```diff
@@ -256,80 +249,98 @@ export const resetFieldForTransaction = (
...
-                action: 'getNewOtp',
-                formId,
-                fieldId,
-                transactionId,
-              },
+                'Error while updating transaction data after sending OTP',
+              meta: logMeta,
               error,
             })
             return new DatabaseError(getMongoErrorMessage(error))
```
Comment (karrui) [inline src/app/modules/verification/verification.service.ts:328]: nit: return `transformMongoError` instead?

File: src/app/modules/verification/__tests__/verification.controller.new.spec.ts [hunk 3]
```diff
@@ -316,4 +331,246 @@ describe('Verification controller', () => {
...
   })
+
+  describe('handleGetOtp', () => {
+    const MOCK_REQ = expressHandler.mockRequest({
+      body: { fieldId: MOCK_FIELD_ID, answer: MOCK_ANSWER },
+      params: { transactionId: MOCK_TRANSACTION_ID },
+    })
+
+    beforeEach(() => {
+      MockOtpUtils.generateOtpWithHash.mockReturnValueOnce(
```
Comment (karrui) [inline src/app/modules/verification/__tests__/verification.controller.new.spec.ts:342]: can set default mock value in a beforeAll
```ts
MockOtpUtils.generateOtpWithHash.mockReturnValue(...)
```

then can just call mockReturnValueOnce in the individual tests to override return value a single time if needed

File: src/app/modules/verification/__tests__/verification.controller.new.spec.ts [hunk 4]
```diff
@@ -316,4 +331,246 @@ describe('Verification controller', () => {
...
+        otp: MOCK_OTP,
+        hashedOtp: MOCK_HASHED_OTP,
+        recipient: MOCK_ANSWER,
+      })
+      expect(mockRes.sendStatus).toHaveBeenCalledWith(StatusCodes.CREATED)
+    })
+
+    it('should return 404 when transaction is not found', async () => {
+      MockVerificationFactory.sendNewOtp
+        .mockReset()
```
Comment (karrui) [inline src/app/modules/verification/__tests__/verification.controller.new.spec.ts:369]: dont need to call mockReset here if default mockReturnValue is given

File: src/app/modules/verification/__tests__/verification.service.new.spec.ts [hunk 5]
```diff
@@ -229,4 +254,246 @@ describe('Verification service', () => {
...
+    let updateHashSpy: jest.SpyInstance<
+      Promise<IVerificationSchema | null>,
+      [updateData: UpdateFieldData]
+    >
+
+    beforeEach(() => {
+      updateHashSpy = jest
+        .spyOn(VerificationModel, 'updateHashForField')
+        .mockResolvedValueOnce(mockTransaction)
+      MockSmsFactory.sendVerificationOtp.mockReturnValueOnce(okAsync(true))
```
Comment (karrui) [inline src/app/modules/verification/__tests__/verification.service.new.spec.ts:268]: Same here, can just do a mockReturnValue in a beforeAll (unless they will get reset somehow)

File: src/app/modules/verification/__tests__/verification.service.new.spec.ts [hunk 6]
```diff
@@ -229,4 +254,246 @@ describe('Verification service', () => {
...
+      ).not.toHaveBeenCalled()
+      expect(updateHashSpy).not.toHaveBeenCalled()
+      expect(result._unsafeUnwrapErr()).toEqual(new TransactionNotFoundError())
+    })
+
+    it('should return TransactionExpiredError when transaction has expired', async () => {
+      const expiredTransaction = await VerificationModel.create({
+        formId: mockFormId,
+        // Expire 25 hours ago
+        expireAt: new Date(Date.now() - 25 * 60 * 60 * 1000),
```
Comment (karrui) [inline src/app/modules/verification/__tests__/verification.service.new.spec.ts:330]: nit: can use `date-fn`s for clarity

**Review by karrui — APPROVED** [review]
lgtm with nits

File: src/app/modules/verification/verification.controller.ts (hunk 1 above)
Comment (mantariksh) [reply src/app/modules/verification/verification.controller.ts:124]: will make the changes in #1455

File: src/app/modules/verification/verification.service.ts (hunk 2 above)
Comment (mantariksh) [reply src/app/modules/verification/verification.service.ts:328]: ok seeing as `DatabaseError` is used everywhere, will make the changes in #1455 to apply this consistently

File: src/app/modules/verification/__tests__/verification.controller.new.spec.ts (hunk 3 above)
Comment (mantariksh) [reply src/app/modules/verification/__tests__/verification.controller.new.spec.ts:342]: will make the changes in #1455

File: src/app/modules/verification/__tests__/verification.service.new.spec.ts (hunk 6 above)
Comment (mantariksh) [reply src/app/modules/verification/__tests__/verification.service.new.spec.ts:330]: updating #1455

File: src/app/modules/verification/verification.controller.ts (hunk 1 above)
Comment (mantariksh) [reply src/app/modules/verification/verification.controller.ts:124]: done

**Comment by mantariksh** [comment]
TODO from previous PRs
- [ ] Use ErrorDto type
- [ ] Use `transformMongoError`
- [ ] Set default mock value in beforeAll instead of using mockReset
- [ ] Use date-fns

**Pull request merged**